clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -809,6 +810,13 @@ private void maybeCloseOldestConnection(long currentTimeNanos) {
      * poll().
      */
     public void clearCompletedReceives() {
+        this.completedReceives.values().forEach(networkReceive -> {
+            try {
+                networkReceive.close();
```

Comment:
Can we add a test for this please?

Reply:
I don't have a good idea on how to assert in the tests that the `NetworkReceive` objects are being closed. I have considered the following:
1. Using `GarbageCollectedMemoryPool` which could have hinted towards a memory leak but it doesn't work.
2. Assert that the NetworkReceive.Close() is being called in the flow but we would need to Spy the objects in the test which is difficult to do here.

I am going to try again on Monday to see if I can figure something out but meanwhile, do you have any suggestions here on how can I add a test to validate the leak?

Reply:
ok, so I was able to write a test that existing code fails by adding a method in `GarbageCollectedMemoryPool` which validates that all buffer has been released. But the memory pool in a selector is not used by anything other than tests. So, production won't have a memory leak here.

The ownership of lifecycle of `NetworkReceive` is confusing and not well defined in the current code. Currently, the list of `CompletedNetworkReceive` is cleared on every poll event (without closing the `NetworkReceive`). Before the poll, a potential caller can obtain reference to `NetworkReceive` using the `Selector.completedReceives()` method. If we close the `NetworkReceive`, the reference available with the caller will not be valid. Which begs the question, is the Selector responsible for closing completed `NetworkReceive(s)` or is it the caller of `Selector.completedReceives()`. The test `SelectorTest.testMuteOnOOM` relies on the fact that caller is responsible for closing the `NetworkReceive`. Although, personally, that doesn't look right to me but nevertheless I don't want to have a discussion about that in the scope of this PR. Since MemoryPool backed `NetworkReceive` is not used by any non-test classes, I am reverting my changes to this class from this PR and will open a separate JIRA ticket to have the discussion about lifecycle ownership.

I hope the above makes sense?

Reply:
Yes, that makes sense. I was unsure how we could have such a bad leak that had gone undetected. It makes sense that the Selector cannot be responsible for closing the network receive since it transfers the ownership to the caller of `poll`. Perhaps we can add a comment explaining this?

Reply:
Thanks. I added the following comment to the `Selector.completedReceives()` in the latest commit.
```
Note that the caller of this method assumes responsibility to close the NetworkReceive resources which may be
backed by a {@link MemoryPool}. In such scenarios (when NetworkReceive uses a {@link MemoryPool}, it is necessary
to close the {@link NetworkReceive} to prevent any memory leaks.
```

The new revision is ready for your review.

Reply:
This code change has been reverted from this code review. Resolving the comment.